### PR TITLE
Refactor "service" registry to use standard REST storage (and generic etcd)

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #12207

### CODE REVIEW

**Comment by wojtek-t** [comment]
@nikhiljindal - I found we will be a good reviewer for it. But feel free to reassign it.

**Comment by wojtek-t** [comment]
@k8s-bot test this please

**Comment by wojtek-t** [comment]
@nikhiljindal FYI: this is blocking #7372

cc @derekwaynecarr

File: pkg/registry/service/etcd/etcd.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/storage"
+)
+
+type REST struct {
+	etcdgeneric.Etcd
+}
+
+func NewStorage(s storage.Interface) *REST {
+	prefix := "/services/specs"
```
Comment (nikhiljindal) [inline pkg/registry/service/etcd/etcd.go]: I know you are not changing this, but do you know why is this not "/services"?

Comment (wojtek-t) [inline pkg/registry/service/etcd/etcd.go]: I have no idea. However, I think that changing it is backward incompatible change, so we can't do it.

**Comment by wojtek-t** [comment]
@nikhiljindal - if you're fine with this PR, can you please apply LGTM label?

File: pkg/registry/service/registry_test.go [hunk 2]
```diff
@@ -236,25 +239,22 @@ func TestEtcdDeleteService(t *testing.T) {
 		t.Errorf("unexpected error: %v", err)
 	}
 
-	if len(fakeClient.DeletedKeys) != 2 {
+	if len(fakeClient.DeletedKeys) != 1 {
```
Comment (nikhiljindal) [inline pkg/registry/service/registry_test.go]: Why is this changing?

File: pkg/registry/service/rest.go [hunk 3]
```diff
@@ -144,6 +144,13 @@ func (rs *REST) Delete(ctx api.Context, id string) (runtime.Object, error) {
 		return nil, err
 	}
 
+	// TODO: can leave dangling endpoints, and potentially return incorrect
```
Comment (nikhiljindal) [inline pkg/registry/service/rest.go]: Why do we want to leave dangling endpoints?

Comment (nikhiljindal) [inline pkg/registry/service/rest.go]: Is the TODO stale since you are deleting the endpoints now?

File: pkg/registry/service/registry_test.go [hunk 4]
```diff
@@ -316,89 +319,6 @@ func TestEtcdWatchServices(t *testing.T) {
 	watching.Stop()
 }
 
-func TestEtcdWatchServicesBadSelector(t *testing.T) {
```
Comment (nikhiljindal) [inline pkg/registry/service/registry_test.go]: Why are we deleting these tests?

**Comment by nikhiljindal** [comment]
Sorry I had started reviewing it but then left it midway. 
Took a complete pass now. Have added a few questions.

File: pkg/registry/service/rest.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/registry/service/rest.go]: I separated this PR into 2 commits.
Basically, the first commit is moving some logic (that is not related to storage operations)
from: pkg/registry/etcd/etcd_test.go
to: pkg/registry/service/rest.go

So I'm not adding it - I'm just moving it to a better location (this is just refactoring, not changing logic).

File: pkg/registry/service/registry_test.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/registry/service/registry_test.go]: As explained above - see separation into 2 commits for explanation for what is happening here (he first commit is what explains it).

File: pkg/registry/service/registry_test.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/registry/service/registry_test.go]: See the first commit of this PR.
Basically, I'm moving "deleting endpoints" from etcd.go into rest.go files. Thus, we longer delete endpoints in the registry (the second key was deleting endpoints). This is now done in rest.go.

**Comment by wojtek-t** [comment]
@nikhiljindal - Thanks a lot for "over-the-weekend" review.

I replied to your comments - PTAL

File: pkg/registry/service/registry_test.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/registry/service/registry_test.go]: ok

**Comment by nikhiljindal** [comment]
Thanks for the clarification.

LGTM.

**Comment by wojtek-t** [comment]
Close & open to trigger shippable.
